R/bayesiannetworkanalysis.R

```diff
@@ -0,0 +1,1297 @@
+    
+    #.networkAnalysisSaveLayout(mainContainer, options, networkList[["layout"]])
+  }
+  
+  return(networkList)
+}
+
+.bayesianNetworkAnalysisComputeCentrality <- function(networks) {
+  
+  centralities <- vector("list", length(networks))
+  for (nw in seq_along(networks)) {
+    
+    network <- networks[[nw]]
+    
+    # add measures 
+    centralitySamples <- centrality(network = network)
+    
+    nSamples <- nrow(network$samplesPosterior)
+    
+    value <- apply(centralitySamples[, 3:(nSamples+2)], MARGIN = 1, mean)
+    centralityHDIintervals <- apply(centralitySamples[, 3:(nSamples+2)], MARGIN = 1, 
+                                    FUN = HDInterval::hdi, allowSplit = FALSE)
+    
+    centralitySummary <- cbind(centralitySamples[, 1:2], value, t(centralityHDIintervals))
+    
+    centralities[[nw]] <- centralitySummary
+
+  }
+  
+  return(centralities)
+}
+
+.bayesianNetworkAnalysisComputeLayout <- function(networks, dataset, options) {
+
+  # Reformat networks to fit averageLayout:
+  weightMatrices <- list()
+  for (i in 1:length(networks)) {
+    weightMatrices[[i]] <- networks[[i]]$graph
+  }
+  jaspBase::.suppressGrDevice(layout <- qgraph::averageLayout(weightMatrices, layout = options[["layout"]], repulsion = options[["repulsion"]]))
+  rownames(layout) <- .unv(colnames(networks[[1L]]))
+
+  return(layout)
+}
+
+.bayesianNetworkAnalysisComputeNetworks <- function(options, dataset) {
+  
+  networks <- vector("list", length(dataset))
+  
+  # For every dataset (given the splitting variable) estimate a network:
+  for (nw in seq_along(dataset)) {
+
+    # When method = "gcgm" a vector with binary values is needed:
+    nonContVariables <- NULL 
+    if (options[["estimator"]] == "gcgm") {
+      nonContVariables <- c()
+      for (var in options[["variables"]]) {
+        
+        # A 1 indicates noncontinuous variables:
+        if (length(levels(factor(dataset[[nw]][[var]]))) <= 8) {
+          nonContVariables <- c(nonContVariables, 1)
+        } else {
+          nonContVariables <- c(nonContVariables, 0)
+        }
+      }
+    }
+    
+    # Check if set.seed needs to be used: 
+    if (options$setSeed) set.seed(options$seed)
+    
+    # Estimate network:  
+    bdgraphFit <- BDgraph::bdgraph(data        = as.data.frame(dataset[[nw]]),
+                                    method     = options[["estimator"]],
+                                    not.cont   = nonContVariables, 
+                                    algorithm  = "rjmcmc",
+                                    iter       = as.numeric(options[["iter"]]),
+                                    save       = TRUE,
+                                    burnin     = as.numeric(options[["burnin"]]),
+                                    g.start    = options[["initialConfiguration"]],
+                                    df.prior   = as.numeric(options[["dfprior"]]),
+                                    g.prior    = as.numeric(options[["gprior"]]))
```

Comment:
If bdgraph can fail and we could show a reasonable error message as to why it failed, it needs a `try`.